tests/contrib/langchain/test_langchain.py

```diff
@@ -1380,39 +1443,139 @@ def test_llmobs_huggingfacehub_llm(self, langchain, mock_llmobs_writer, mock_tra
             huggingfacehub_api_token=os.getenv("HUGGINGFACEHUB_API_TOKEN", "<not-a-real-key>"),
         )
 
-        self._test_llmobs_invoke(
-            provider="huggingface_hub",
+        self._test_llmobs_llm_invoke(
             generate_trace=llm,
             request_vcr=request_vcr,
             mock_llmobs_writer=mock_llmobs_writer,
             mock_tracer=mock_tracer,
             cassette_name="huggingfacehub_completion_sync.yaml",
+            provider="huggingface_hub",
         )
 
     def test_llmobs_openai_chat_model(self, langchain, mock_llmobs_writer, mock_tracer, request_vcr):
         chat = langchain.chat_models.ChatOpenAI(temperature=0, max_tokens=256)
 
-        self._test_llmobs_invoke(
-            provider="openai",
+        self._test_llmobs_llm_invoke(
             generate_trace=lambda prompt: chat([langchain.schema.HumanMessage(content=prompt)]),
             request_vcr=request_vcr,
             mock_llmobs_writer=mock_llmobs_writer,
             mock_tracer=mock_tracer,
             cassette_name="openai_chat_completion_sync_call.yaml",
+            provider="openai",
             input_role="user",
             output_role="assistant",
+            different_py39_cassette=True,
         )
 
     def test_llmobs_openai_chat_model_custom_role(self, langchain, mock_llmobs_writer, mock_tracer, request_vcr):
         chat = langchain.chat_models.ChatOpenAI(temperature=0, max_tokens=256)
 
-        self._test_llmobs_invoke(
-            provider="openai",
+        self._test_llmobs_llm_invoke(
             generate_trace=lambda prompt: chat([langchain.schema.ChatMessage(content=prompt, role="custom")]),
             request_vcr=request_vcr,
             mock_llmobs_writer=mock_llmobs_writer,
             mock_tracer=mock_tracer,
             cassette_name="openai_chat_completion_sync_call.yaml",
+            provider="openai",
             input_role="custom",
             output_role="assistant",
+            different_py39_cassette=True,
+        )
+
+    def test_llmobs_chain(self, langchain, mock_llmobs_writer, mock_tracer, request_vcr):
+        chain = langchain.chains.LLMMathChain(llm=langchain.llms.OpenAI(temperature=0, max_tokens=256))
+
+        self._test_llmobs_chain_invoke(
+            generate_trace=lambda prompt: chain.run("what is two raised to the fifty-fourth power?"),
+            request_vcr=request_vcr,
+            mock_llmobs_writer=mock_llmobs_writer,
+            mock_tracer=mock_tracer,
+            cassette_name="openai_math_chain_sync.yaml",
+            expected_spans_data=[
+                (
+                    "chain",
+                    {
+                        "input_value": str({"question": "what is two raised to the fifty-fourth power?"}),
+                        "output_value": mock.ANY,
```

Comment:
Are the outputs deterministic? Can we specify the output values instead of relying on `mock.ANY`? (Same for the other LLMObs chain tests)

Reply:
yeah I can. do we want to do this for _all_ of these chaining tests, or maybe just the basic `test_llmobs_chain` ones? some of the outputs are a bit longer (from the cassettes), and since we don't truncate for LLMObs spans, I just thought it would be less clutter this way, and make it easier to read. But I do agree that we should have it for each basic chain test in each suite

Reply:
Good point, let's add the assertions for each basic chain test just as a sanity check. As long as we have coverage I'm good with that

Reply:
added. one of them is marked as flaky, but I made sure it (x)passed before committing 🥳